includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -975,13 +985,14 @@ public function ajax_get_selected_product_data() {
 		check_ajax_referer( 'wcpay-get-selected-product-data', 'security' );
 
 		try {
-			$product_id   = isset( $_POST['product_id'] ) ? absint( $_POST['product_id'] ) : false;
-			$qty          = ! isset( $_POST['qty'] ) ? 1 : apply_filters( 'woocommerce_add_to_cart_quantity', absint( $_POST['qty'] ), $product_id );
-			$addon_value  = isset( $_POST['addon_value'] ) ? max( (float) $_POST['addon_value'], 0 ) : 0;
-			$product      = wc_get_product( $product_id );
-			$variation_id = null;
-			$currency     = get_woocommerce_currency();
-			$is_deposit   = isset( $_POST['wc_deposit_option'] ) ? 'yes' === sanitize_text_field( wp_unslash( $_POST['wc_deposit_option'] ) ) : null;
+			$product_id      = isset( $_POST['product_id'] ) ? absint( $_POST['product_id'] ) : false;
+			$qty             = ! isset( $_POST['qty'] ) ? 1 : apply_filters( 'woocommerce_add_to_cart_quantity', absint( $_POST['qty'] ), $product_id );
+			$addon_value     = isset( $_POST['addon_value'] ) ? max( (float) $_POST['addon_value'], 0 ) : 0;
+			$product         = wc_get_product( $product_id );
+			$variation_id    = null;
+			$currency        = get_woocommerce_currency();
+			$is_deposit      = isset( $_POST['wc_deposit_option'] ) ? 'yes' === sanitize_text_field( wp_unslash( $_POST['wc_deposit_option'] ) ) : null;
+			$deposit_plan_id = isset( $_POST['wc_deposit_payment_plan'] ) ? absint( $_POST['wc_deposit_payment_plan'] ) : 0;
```

Comment:
Depending on the outcome of my other comment, we might remove this parameter (because it's currently only used to check if a deposit plan is specified as `0 === $deposit_plan_id` and is part of the logic I suggest to moving to deposits), and let Deposits get it from the request.

Otherwise, it looks good with a caveat that we could add a layer of security by verifying if a deposit plan with `$deposit_plan_id` exists in the system. This would require Deposits with the method like, e.g., `WC_Deposits_Plan::exists( $deposit_plan_id )` if we'll keep this parameter in our code, alternatively if we address my comment then the verification will occur on the Deposits side (I think [currently it might fail for a plan which doesn't exist](https://github.com/woocommerce/woocommerce-deposits/blob/trunk/includes/class-wc-deposits-plan.php#L51-L57))

Reply:
> because it's currently only used to check if a deposit plan is specified as 0 === $deposit_plan_id

When it is not `0`, we use the value. We only check if it is `0` to overwrite it.

> Otherwise, it looks good with a caveat that we could add a layer of security by verifying if a deposit plan with $deposit_plan_id exists in the system. 

I don't think checking if it exists is a security layer. We would just be replacing an unhandled error from Deposits with a handled error from our side, but the result would be the same and only displayed to someone who is trying to hack the system.
That said, I agree with checking if the plan is available for the product. Someone could buy a product that only accepts Plan 1 and Plan 2 using Plan 3, and I don't think anything is preventing them from doing that in `get_deposit_amount`, so I'm adding a check for that.

Reply:
Thanks for mentioning in another comment that getting request params should be done here rather than further during the process: this almost makes my initial message in this thread irrelevant. Also, thanks for adding the check for plan availability for the product. 👍 I'm closing this thread.